### PR TITLE
Add mixed precision

### PR DESCRIPTION
**Pull request by andersonic**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Add mixed precision training to Adam. Update benchmark.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: benchmarks/transformer.py [hunk 1]
```diff
@@ -221,7 +223,7 @@ def benchmark_language_model(train_data, val_data, test_data, model, criterion,
     if can_benchmark and len(model.balance) == 4:
         # Assert that words per second is within 3 standard deviations of the average
         # of six golden runs
-        assert wps > 20052.1 - (3 * 359)
+        assert wps > 27799.2 - (3 * 522.145)
```
Comment (blefaudeux) [inline benchmarks/transformer.py:226]: nice speedups and memory reduction (below) !

Comment (andersonic) [reply benchmarks/transformer.py:226]: Thanks!

File: benchmarks/transformer.py [hunk 2]
```diff
@@ -230,10 +232,10 @@ def benchmark_language_model(train_data, val_data, test_data, model, criterion,
 
         # Assert that memory usage on each GPU is within 10% of golden run
         # Right-hand-side is golden run bytes * 110%
-        assert torch.cuda.memory_stats(0)["allocated_bytes.all.peak"] < 365916160 * 1.1
-        assert torch.cuda.memory_stats(1)["allocated_bytes.all.peak"] < 1281024 * 1.1
-        assert torch.cuda.memory_stats(2)["allocated_bytes.all.peak"] < 2788864 * 1.1
-        assert torch.cuda.memory_stats(3)["allocated_bytes.all.peak"] < 190724608 * 1.1
+        assert torch.cuda.memory_stats(0)["allocated_bytes.all.peak"] < 210479616 * 1.1
```
Comment (sidgoyal78) [inline benchmarks/transformer.py:235]: Nice, i'm curious what tool did you use get these exact numbers?

File: fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu [hunk 3]
```diff
@@ -40,26 +39,26 @@ struct AdamFunctor
...
         int n = tl.sizes[tensor_loc];
 
-        GRAD_T* p = (GRAD_T *)tl.addresses[0][tensor_loc];
+        PARAM_T* p = (PARAM_T *)tl.addresses[0][tensor_loc];
         p += chunk_idx*chunk_size;
-        T* m = (T *)tl.addresses[1][tensor_loc];
+        float* m = (float *)tl.addresses[1][tensor_loc];
         m += chunk_idx*chunk_size;
-        T* v = (T *)tl.addresses[2][tensor_loc];
+        float* v = (float *)tl.addresses[2][tensor_loc];
```
Comment (sidgoyal78) [inline fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu:46]: naive question: why do we have types of m and v as float?

File: fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu [hunk 4]
```diff
@@ -135,24 +134,62 @@ void fused_adam_cuda(
...
+            AdamFunctor<5, float, at::Half>(),
+            beta1,
+            beta2,
+            eps,
+            grad_scale,
+            step_size,
+            (adamMode_t) mode,
+            decay
+        );
+    } else {    // tl_sz == 4
```
Comment (sidgoyal78) [inline fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu:158]: maybe add an explicit comment for the setting here, similar to "mix precision case" for the above :)

File: benchmarks/transformer.py (hunk 2 above)
Comment (andersonic) [reply benchmarks/transformer.py:235]: I used the values printed by the above four lines of code

File: fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu (hunk 3 above)
Comment (andersonic) [reply fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu:46]: These are momentum and velocity! And right now, we require them to be floats; in the Python code, when they are instantiated, they are always dtype=torch.float32. Next pull request will add in the option for them to be fp16

File: fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu (hunk 4 above)
Comment (andersonic) [reply fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu]: Great idea! Done!

**Pull request merged**